pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -1184,7 +1184,6 @@ protected void handleSubscribe(final CommandSubscribe subscribe) {
                         ServerError error = getErrorCodeWithErrorLog(existingConsumerFuture, true,
                                 String.format("Consumer subscribe failure. remoteAddress: %s, subscription: %s",
                                         remoteAddress, subscriptionName));
-                        consumers.remove(consumerId, existingConsumerFuture);
```

Comment:
There are three clients in your example, the pulsar client maintains its own connection pool, which means the `client1` will not use the same connection as `client2` or `client3`. 

The code you changed is the code of `ServerCnx`(this object is not shared across multi clients), so it will not affect other consumers of other clients

Reply:
Additional background: 

If you have two pulsar clients, the consumers registered will be like this:
- pulsar-client-1
  - connection-1 (one-to-one relationship with `ServerCnx`)
    - consumer-1
    - consumer-2 
  - connection-2
    - consumer-3
    - consumer-4 
- pulsar-client-2
  - connection-3
    - consumer-5
    - consumer-6 
  - connection-4  
    - consumer-7
    - consumer-8

Reply:
So only `consumer-1` and `consumer-2` can be in conflict since their `consumer-id` is not the same, so all things are OK.

Reply:
I guess it is one client request three times. Because in our server log, "Subscribing on topic" occur three times in a short time from the same remoteAddress host:ip. Therefore they are in the same connection ?

Reply:
> I guess it is one client request three times. Because in our server log, "Subscribing on topic" occur three times in a short time from the same remoteAddress host:ip. Therefore they are in the same connection ?

could you provide the logs?

Reply:
I guess the three request is from the same client, handleSubscribe the same consumerId and same remoteAddress three times. 

```
11:35:42.794 [pulsar-io-4-39] INFO  org.apache.pulsar.broker.service.ServerCnx - [/host1:ip1] Subscribing on topic persistent://xxx/xxx/topic-partition-3 / subscription-xxx-1

11:36:12.793 [pulsar-io-4-39] INFO  org.apache.pulsar.broker.service.ServerCnx - [/host1:ip1] Closing consumer: consumerId=3
11:36:12.793 [pulsar-io-4-39] INFO  org.apache.pulsar.broker.service.ServerCnx - [/host1:ip1] Closed consumer before its creation was completed. consumerId=3

11:36:12.990 [pulsar-io-4-39] INFO  org.apache.pulsar.broker.service.ServerCnx - [/host1:ip1] Subscribing on topic persistent://xxx/xxx/topic-partition-3 / subscription-xxx-1
11:36:12.990 [pulsar-io-4-39] ERROR org.apache.pulsar.broker.service.ServerCnx - Consumer subscribe failure. remoteAddress: /host1:ip1, subscription: subscription-xxx-1
java.util.concurrent.CompletionException: java.lang.IllegalStateException: Closed consumer before creation was complete

11:36:13.389 [pulsar-io-4-39] INFO  org.apache.pulsar.broker.service.ServerCnx - [/host1:ip1] Subscribing on topic persistent://xxx/xxx/topic-partition-3 / subscription-xxx-1

11:36:13.685 [broker-topic-workers-OrderedScheduler-1-0] INFO  org.apache.pulsar.broker.service.ServerCnx - [/host1:ip1] Created subscription on topic persistent://xxx/xxx/topic-partition-3 / subscription-xxx-1


11:36:13.706 [broker-topic-workers-OrderedScheduler-1-0] INFO  org.apache.pulsar.broker.service.persistent.PersistentDispatcherMultipleConsumers - Removed consumer Consumer{subscription=PersistentSubscription{topic=persistent://xxx/xxx/topic-partition-3, name=subscription-xxx-1}, consumerId=3, consumerName=2b83e9380c, address=/host1:ip1} with pending 0 acks

11:36:13.706 [broker-topic-workers-OrderedScheduler-1-0] INFO  org.apache.pulsar.broker.service.ServerCnx - [/host1:ip1] Cleared consumer created after timeout on client side Consumer{subscription=PersistentSubscription{topic=persistent://xxx/xxx/topic-partition-3, name=subscription-xxx-1}, consumerId=3, consumerName=2b83e9380c, address=/host1:ip1}
```

Reply:
Maybe the picture is somewhat misunderstanding. I have modified it.

Reply:
By the log, we can not confirm if these three consumers are the same, because we can not know their `consumer id`. This log was been improved in the PR https://github.com/apache/pulsar/pull/20568/files#diff-1e0e8195fb5ec5a6d79acbc7d859c025a9b711f94e6ab37c94439e99b3202e84R1168

Reply:
| time | `req-1` | `req-2` | `req-3` | `close connection` | 
| --- | --- | --- | --- | --- |
| 1 | | | put a new `consumerFuture` |
| 2 | | | | close connection |
| 3 | | | | mark the `consumerFuture` as failed |
| 4 | | got `consumerFuture`(failed) | 
| 5 | | remove |
| 6 | | | add the consumer into the list and set(`list.size = 1` and `set.size = 1`) |
| 7 | put the second `consumerFuture` | |
| 8 | add the second consumer into the list and set(`list.size = 2` and `set.size = 1`) |
| 7 | | | remove the consumer from the list and set(`list.size = 1` and `set.size = 0`) |

Do you want to say the three requests executed as above?

Reply:
BTW, there is a PR trying to fix the concurrent call `subscribe` in the same client 
- https://github.com/apache/pulsar/pull/20591

I still think the current PR is trying to solve this issue is meaningful, and it would be nice to have a test that can reproduce it.

Reply:
@TakaHiR07 

Could you take a look at this PR https://github.com/apache/pulsar/pull/22283 and modify this one? If yes, I will close https://github.com/apache/pulsar/pull/22283 ❤️